### PR TITLE
ROC-3252 Add fallback for bulk print

### PR DESCRIPTION
**Pull request by dharmendrak**
https://dev-build.reform.hmcts.net/blue/organizations/jenkins/hmcts%2Fcmc-claim-store/detail/PR-248/26/pipeline/

### CODE REVIEW

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@TestPropertySource(
+    properties = {
+        "document_management.api_gateway.url=false",
+        "core_case_data.api.url=false",
+        "send-letter.url=http://localhost:8089/send"
+    }
+)
+@AutoConfigureWireMock(port = 8089)
```
Comment (rlewan) [inline src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java]: Please don't use a fixed port, there is no guarantee this port will be available on the Jenkins machine. Wiremock should provide a feature to determine an open port by itself.

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                .withStatus(HttpStatus.BAD_REQUEST.value())
+                .withHeader("Content-Type", "application/json")
+            )
+        );
+
+        MvcResult result = makeRequest(SampleClaimData.submittedByClaimant())
+            .andExpect(status().isOk())
+            .andReturn();
+
+        verify(bulkPrintNotificationService, atLeast(1))
```
Comment (rlewan) [inline src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java]: Why do we use `atLeast` one here? Shouldn't there be only one notification expected?

Comment (rlewan) [reply src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java]: Or 3 as the test name suggests?

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            .willReturn(aResponse()
+                .withStatus(HttpStatus.INTERNAL_SERVER_ERROR.value())
+                .withHeader("Content-Type", "application/json")
+                .withBody("Internal server error occurred")));
+
+        MvcResult result = makeRequest(SampleClaimData.submittedByClaimant())
+            .andExpect(status().isOk())
+            .andReturn();
+
+        verify(bulkPrintNotificationService, atLeast(1))
```
Comment (rlewan) [inline src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java]: Same here.

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            .builder()
+            .build();
+
+        when(pdfServiceClient.generateFromHtml(any(byte[].class), anyMap()))
+            .thenReturn(PDF_CONTENT);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNullPointerWhenGivenNullClaim() {
+        service.notifyFailedBulkPrint(null, null);
```
Comment (rlewan) [inline src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationServiceTest.java]: There are 2 `null` values passed in here but the test mentions claim only. Please provide a value for the other one and an additional test to check that NPE is thrown when the other one is `null`.

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        verify(emailService).sendEmail(senderArgument.capture(), emailDataArgument.capture());
+
+        EmailAttachment emailAttachment = emailDataArgument.getValue()
+            .getAttachments()
+            .get(0);
+
+        String expectedPinLetterFileName
+            = DocumentNameUtils.buildDefendantLetterFileBaseName(claim.getReferenceNumber());
+
+        assertThat(emailAttachment.getContentType()).isEqualTo("application/pdf");
```
Comment (rlewan) [inline src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationServiceTest.java]: Can you please check if Spring provides a constant for this content type?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationService.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    public void notifyFailedBulkPrint(List<Document> documents, Claim claim) {
+        EmailContent emailContent = emailContentProvider.createContent(wrapInMap(claim));
+
+        EmailAttachment defendantLetter = pdf(
+            createPdf(documents.get(0)),
+            buildDefendantLetterFileBaseName(claim.getReferenceNumber())
+        );
+
+        EmailAttachment sealedClaim = pdf(
+            createPdf(documents.get(1)),
```
Comment (rlewan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationService.java]: Can you please check if we have an automatic test which verifies that the documents will always appear in this order?

Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationService.java]: there is an unit test

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        MvcResult result = makeRequest(SampleClaimData.submittedByClaimant())
+            .andExpect(status().isOk())
+            .andReturn();
+
+        verify(bulkPrintNotificationService, never())
+            .notifyFailedBulkPrint(any(List.class), eq(deserializeObjectFrom(result, Claim.class)));
+    }
+
+    @Test
+    public void shouldTrySendingLetterThreeTimesOnFailuresWhenBulkPrintFailsWithSendLetterException() throws Exception {
```
Comment (geek-soft) [inline src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java]: Please rename, test does not check retry feature at all, just handling failure.

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        MvcResult result = makeRequest(SampleClaimData.submittedByClaimant())
+            .andExpect(status().isOk())
+            .andReturn();
+
+        verify(bulkPrintNotificationService)
+            .notifyFailedBulkPrint(any(List.class), eq(deserializeObjectFrom(result, Claim.class)));
+    }
+
+    @Test
+    public void shouldTrySendingLetterThreeTimesOnFailuresWhenBulkPrintFailsWithRestClientException() throws Exception {
```
Comment (geek-soft) [inline src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java]: The same here please

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+        when(authTokenGenerator.generate()).thenReturn(AUTHORISATION_TOKEN);
+
```
Comment (geek-soft) [inline src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java]: Remove blank line for consistency.

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        wireMockServer.stubFor(post(urlEqualTo("/send/letters"))
+            .willReturn(aResponse()
+                .withStatus(HttpStatus.INTERNAL_SERVER_ERROR.value())
```
Comment (geek-soft) [inline src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java]: Can you explain the difference in the way application handles 400 response and 500 response? I am trying to see why we have two test cases simulating these two response code.

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    public void shouldSendEmailWithExpectedPDFAttachments() throws IOException {
+        service.notifyFailedBulkPrint(documents, claim);
+
+        verify(emailService).sendEmail(senderArgument.capture(), emailDataArgument.capture());
+
+        EmailAttachment emailAttachment = emailDataArgument.getValue()
+            .getAttachments()
+            .get(0);
+
+        String expectedPinLetterFileName
```
Comment (geek-soft) [inline src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationServiceTest.java]: That test does not guarantee that in real life first document will be a PIN letter and second one will be sealed claim copy. 

I can change `publisher.publishEvent(new DocumentReadyToPrintEvent(event.getClaim(), defendantLetterDocument , sealedClaimDocument));` to be` publisher.publishEvent(new DocumentReadyToPrintEvent(event.getClaim(), sealedClaimDocument, defendantLetterDocument));` and tests will keep passing while in application will start sending PIN letters as sealed claims and vice versa.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationService.java (hunk 6 above)
Comment (geek-soft) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationService.java]: The test is useless in that matter, refactoring can easily bring breaking change that is not detected. See my comment above.

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationServiceTest.java (hunk 11 above)
Comment (dharmendrak) [reply src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationServiceTest.java]: BulkPrintServiceTest checks for the order of documents and will fail if order changes

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/MockSpringTest.java [hunk 12]
```diff
@@ -23,11 +23,10 @@
 import uk.gov.hmcts.reform.document.DocumentMetadataDownloadClientApi;
 import uk.gov.hmcts.reform.document.DocumentUploadClientApi;
 import uk.gov.hmcts.reform.pdf.service.client.PDFServiceClient;
-import uk.gov.hmcts.reform.sendletter.api.SendLetterApi;
 import uk.gov.service.notify.NotificationClient;
 
 @RunWith(SpringRunner.class)
-@SpringBootTest
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
```
Comment (rlewan) [inline src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/MockSpringTest.java]: I think that's not needed, Spring does not need to open a port, only Wiremock does.

Comment (dharmendrak) [reply src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/MockSpringTest.java]: you are right, it works fine without it, however https://cloud.spring.io/spring-cloud-contract/1.1.x/multi/multi__spring_cloud_contract_wiremock.html does say to use it.

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java (hunk 10 above)
Comment (dharmendrak) [reply src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java]: it has no purpose as it is handling both exceptions in same way. However wanted to keep two exceptions in sendLetterApi, if other user want to handle them separately.

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java (hunk 7 above)
Comment (dharmendrak) [reply src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java]: done

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java (hunk 8 above)
Comment (dharmendrak) [reply src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java]: done

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java (hunk 9 above)
Comment (dharmendrak) [reply src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/controllers/BulkPrintRequestTest.java]: done

File: src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationServiceTest.java (hunk 11 above)
Comment (geek-soft) [reply src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/services/staff/BulkPrintStaffNotificationServiceTest.java]: No it does not, it checks only that the order is not changed within the `BulkPrintServiceTest`

**Comment by geek-soft** [comment]
@dharmendrak please change `DocumentReadyToPrintEvent` to have two properties (defendantLetterDocument and sealedClaimDocument) instead of using `List<Document> documents`. That will guarantee ordering. Then in unit tests make sure letter goes first followed by sealed claim.

**Pull request merged**